### PR TITLE
Refactor SteadystateProblem triple nested try catch clause

### PR DESCRIPTION
**Pull request by paulstapor**
This should at least resolve the nested try catch and reduce the overall complexity of the steady state solver.

### CODE REVIEW

File: src/steadystateproblem.cpp [hunk 1]
```diff
@@ -148,6 +83,89 @@ void SteadystateProblem::workSteadyStateProblem(Solver *solver, Model *model,
                                     SensitivityOrder::first);
 }
 
+NewtonStatus SteadystateProblem::findSteadyState(Solver *solver,
+                                                 NewtonSolver *newtonSolver,
+                                                 Model *model, int it) {
+    /* First, try to run the Newton solver */
```
Comment (FFroehlich) [inline src/steadystateproblem.cpp:89]: could we flatten/deduplicate this further by introducing `findSteadyStateNewton` and `findSteadyStateSimulation`  where exception behaviour is modulated by a flag which is only passed on the second call to `findSteadyStateNewton`?

Would also be nice if we wouldn't have to have duplicate code to copy of the number of steps. Could be implemented by a context manager which is either explicitly in the exception or goes out of context when the status is returned, see https://github.com/ICB-DCM/AMICI/blob/ffb5f33001047556a718c504ad17bc81ed3e206d/include/amici/forwardproblem.h as example

Comment (dweindl) [reply src/steadystateproblem.cpp:89]: > context manager

Wouldn't use too many of those. In python they are fine, but in C++ usually we don't want to have side effects in destructors. Especially destructors should never `throw`.

File: src/steadystateproblem.cpp [hunk 2]
```diff
@@ -83,30 +84,67 @@ void SteadystateProblem::workSteadyStateProblem(Solver *solver, Model *model,
...
+                steady_state_status[0] = SteadyStateStatus::failed_damping;
+                break;
+            default:
+                steady_state_status[0] = SteadyStateStatus::failed;
+        }
+    }
+
+    /* copy number of linear steps used */
+    if (maxSteps > 0)
+        std::copy_n(newtonSolver->getNumLinSteps().begin(),
```
Comment (FFroehlich) [inline src/steadystateproblem.cpp:134]: I believe this needs a check for `newton_retry` to switch between `numlinsteps.begin()` and `&numlinsteps.at(maxSteps)`

File: src/rdata.cpp [hunk 3]
```diff
@@ -188,11 +190,11 @@ void ReturnData::processPreEquilibration(SteadystateProblem const &preeq,
         for (int ip = 0; ip < nplist; ip++)
             writeSlice(sx_rdata[ip], slice(sx_ss, ip, nx));
     }
-    /* Get cpu time for Newton solve in seconds */
-    preeq_cpu_time = preeq.getCPUTime() / 1000;
-    preeq_status = static_cast<int>(preeq.getNewtonStatus());
+    /* Get cpu time for Newton solve in milliseconds */
```
Comment (FFroehlich) [inline src/rdata.cpp:193]: cpu time for simulation is reported in ms though, right?

Comment (paulstapor) [reply src/rdata.cpp:193]: I wanted to have things in the same unit... Now, everything should be in milliseconds, hopefully, but will recheck

File: src/steadystateproblem.cpp (hunk 1 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp:90]: I really appreciate the idea of further flattening out the logic. I think that makes it much more transparent.

File: src/steadystateproblem.cpp (hunk 2 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp:134]: True! Thanks!

File: src/steadystateproblem.cpp [hunk 4]
```diff
@@ -191,9 +262,20 @@ bool SteadystateProblem::checkConvergence(const Solver *solver, Model *model) {
     return converged;
 }
 
+bool SteadystateProblem::checkSteadyStateSuccess() {
+    /* Did one of the attempts yield s steady state? */
+    if (steady_state_status[0] == SteadyStateStatus::success ||
```
Comment (FFroehlich) [inline src/steadystateproblem.cpp:267]: I think using `std::any_of` would be more idiomatic. Should be something like `std::any_of(steady_state_status.begin(), steady_state_status.end(), [](SteadyStateStatus status){return status == SteadyStateStatus::success;})`

May even be shorter but definitely better readability and easier to mantain if we decide to extend `steady_state_status`.

File: src/steadystateproblem.cpp (hunk 1 above)
Comment (FFroehlich) [reply src/steadystateproblem.cpp:90]: Agreed, much cleaner. Pretty happy with the result, thanks for putting in the effort!

Comment (paulstapor) [reply src/steadystateproblem.cpp:90]: Welcome and thanks! :)
As however, we need some extended functionality in this thing (the fully adjoint preequilibration) our large scale project, leaving the code in the 2017-version was not really an option any more... ;)
There will be two more PRs, unfortunately, until then:
1. The fully adjoint preequilibration, first only using the algebraic solution (i.e., solving the linear system of equations), see #1128 .
2. Adding an option to workSteadyStateBackward (then automatically preeq and posteq) to use simulation, including interfacing the cvodes api for the forward simulation quadratures...

However, then preequilibration should have become solid, also for really large scale models with whatever conservation laws...

**Comment by paulstapor** [comment]
Would merge then, if it's fine.

**Pull request merged**